### ISSUE TITLE
Subscription.Cancel() hangs forever if PubSub exited

### ISSUE THREAD

**Issue by lukesolo** [libp2p/go-libp2p-pubsub#200]
Function Cancel() blocks forever because no one listens to sub.cancelCh if PubSub has exited.

```golang
func (sub *Subscription) Cancel() {
	sub.cancelCh <- sub
}
```

How to reproduce:

```golang
package main

import (
	"context"
	"fmt"
	"time"

	"github.com/libp2p/go-libp2p"
	pubsub "github.com/libp2p/go-libp2p-pubsub"
)

func main() {
	ctx, cancel := context.WithCancel(context.Background())

	host, _ := libp2p.New(ctx)
	gsub, _ := pubsub.NewGossipSub(ctx, host)
	sub, _ := gsub.Subscribe("topic")

	cancel()
	time.Sleep(time.Second)

	fmt.Println("cancelling subscription")
	sub.Cancel() // will hang forever
	fmt.Println("subscription cancelled")
}
```

I think that Subscription should have the same context as a GossipSub, so that you could check it in the Cancel function

```golang
func (sub *Subscription) Cancel() {
	select {
		case sub.cancelCh <- sub:
		case <-sub.ctx.Done():
	}
}
```

**Comment by vyzo** [comment]
Yes indeed, this is a bug; care for a patch?

**Comment by lukesolo** [comment]
@vyzo This is not a critical bug for me (at least now), but I made a pull request with a simple fix, and now I think what to do with the other methods which get context as an argument. Should they check both inner and outer context for cancellation in select statement?

**Comment by vyzo** [comment]
If they are doing an operation with the event loop only, because that's only when they can deadlock on shutdown.

**Comment by vyzo** [comment]
Closing as the problem in the issue has been fixed by #201
